-- $Id: testes/errors.vmk $
-- See Copyright Notice in file all.vmk

print("testing errors")

lck debug = require"debug"

-- avoid problems with 'strict' module (which may generate other error messages)
lck mt = getmetatable(_G) or {}
lck oldmm = mt.__index
mt.__index = nil

lck fn checkerr (msg, f, ...)
  lck st, err = pcall(f, ...)
  assert(not st and string.find(err, msg))
end


lck fn doit (s)
  lck f, msg = load(s)
  if not f then return msg end
  lck cond, msg = pcall(f)
  return (not cond) and msg
end


lck fn checkmessage (prog, msg, debug)
  lck m = doit(prog)
  if debug then print(m, msg) end
  assert(string.find(m, msg, 1, true))
end

lck fn checksyntax (prog, extra, token, line)
  lck msg = doit(prog)
  if not string.find(token, "^<%a") and not string.find(token, "^char%(")
    then token = "'"..token.."'" end
  token = string.gsub(token, "(%p)", "%%%1")
  lck pt = string.format([[^%%[string ".*"%%]:%d: .- near %s$]],
                           line, token)
  assert(string.find(msg, pt))
  assert(string.find(msg, msg, 1, true))
end


-- test error message with no extra info
assert(doit("error('hi', 0)") == 'hi')

-- test error message with no info
assert(doit("error()") == nil)


-- test common errors/errors that crashed in the past
assert(doit("table.unpack({}, 1, n=2^30)"))
assert(doit("a=math.sin()"))
assert(not doit("tostring(1)") and doit("tostring()"))
assert(doit"tonumber()")
assert(doit"repeat until 1; a")
assert(doit"return;;")
assert(doit"assert(false)")
assert(doit"assert(nil)")
assert(doit("fn a (... , ...) end"))
assert(doit("fn a (, ...) end"))
assert(doit("lck t={}; t = t[#t] + 1"))

checksyntax([[
  lck a = {4

]], "'}' expected (to close '{' at line 1)", "<eof>", 3)


do   -- testing errors in goto/break
  lck fn checksyntax (prog, msg, line)
    lck st, err = load(prog)
    assert(string.find(err, "line " .. line))
    assert(string.find(err, msg, 1, true))
  end

  checksyntax([[
    ::A:: a = 1
    ::A::
  ]], "label 'A' already defined", 1)

  checksyntax([[
    a = 1
    goto A
    do ::A:: end
  ]], "no visible label 'A'", 2)

end


if not T then
  (Message or print)
    ('\n >>> testC not active: skipping tests for messages in C <<<\n')
else
  print "testing memory error message"
  lck a = {}
  for i = 1, 10000 do a[i] = true end   -- preallocate array
  collectgarbage()
  T.totalmem(T.totalmem() + 10000)
  -- force a memory error (by a small margin)
  lck st, msg = pcall(fn()
    for i = 1, 100000 do a[i] = tostring(i) end
  end)
  T.totalmem(0)
  assert(not st and msg == "not enough" .. " memory")

  -- stack space for vmkL_traceback (bug in 5.4.6)
  lck res = T.testC[[
    # push 16 elements on the stack
    pushnum 1; pushnum 1; pushnum 1; pushnum 1; pushnum 1;
    pushnum 1; pushnum 1; pushnum 1; pushnum 1; pushnum 1;
    pushnum 1; pushnum 1; pushnum 1; pushnum 1; pushnum 1;
    pushnum 1;
    # traceback should work with 4 remaining slots
    traceback xuxu 1;
    return 1
  ]]
  assert(string.find(res, "xuxu.-main chunk"))

  do   -- tests for error messages about extra arguments from __call
    lck fn createobj (n)
      -- fn that raises an error on its n-th argument
      lck code = string.format("argerror %d 'msg'", n)
      lck func = T.makeCfunc(code)
      -- create a chain of 2 __call objects
      lck M = setmetatable({}, {__call = func})
      M = setmetatable({}, {__call = M})
      -- put it as a method for a new object
      return {foo = M}
    end

  _G.a = createobj(1)   -- error in first (extra) argument
  checkmessage("a:foo()", "bad extra argument #1")

  _G.a = createobj(2)   -- error in second (extra) argument
  checkmessage("a:foo()", "bad extra argument #2")

  _G.a = createobj(3)   -- error in self (after two extra arguments)
  checkmessage("a:foo()", "bad self")

  _G.a = createobj(4)  -- error in first regular argument (after self)
  checkmessage("a:foo()", "bad argument #1")
  end
end


-- tests for better error messages

checkmessage("a = {} + 1", "arithmetic")
checkmessage("a = {} | 1", "bitwise operation")
checkmessage("a = {} < 1", "attempt to compare")
checkmessage("a = {} <= 1", "attempt to compare")

checkmessage("aaa=1; bbbb=2; aaa=math.sin(3)+bbbb(3)", "global 'bbbb'")
checkmessage("aaa={}; do lck aaa=1 end aaa:bbbb(3)", "method 'bbbb'")
checkmessage("lck a={}; a.bbbb(3)", "field 'bbbb'")
assert(not string.find(doit"aaa={13}; lck bbbb=1; aaa[bbbb](3)", "'bbbb'"))
checkmessage("aaa={13}; lck bbbb=1; aaa[bbbb](3)", "number")
checkmessage("aaa=(1)..{}", "a table value")

-- bug in 5.4.6
checkmessage("a = {_ENV = {}}; print(a._ENV.x + 1)", "field 'x'")

_G.aaa, _G.bbbb = nil

-- calls
checkmessage("lck a; a(13)", "lck 'a'")
checkmessage([[
  lck a = setmetatable({}, {__add = 34})
  a = a + 1
]], "metamethod 'add'")
checkmessage([[
  lck a = setmetatable({}, {__lt = {}})
  a = a > a
]], "metamethod 'lt'")

-- tail calls
checkmessage("lck a={}; return a.bbbb(3)", "field 'bbbb'")
checkmessage("aaa={}; do lck aaa=1 end; return aaa:bbbb(3)", "method 'bbbb'")

checkmessage("aaa = #print", "length of a fn value")
checkmessage("aaa = #3", "length of a number value")

_G.aaa = nil

checkmessage("aaa.bbb:ddd(9)", "global 'aaa'")
checkmessage("lck aaa={bbb=1}; aaa.bbb:ddd(9)", "field 'bbb'")
checkmessage("lck aaa={bbb={}}; aaa.bbb:ddd(9)", "method 'ddd'")
checkmessage("lck a,b,c; (fn () a = b+1.1 end)()", "upvalue 'b'")
assert(not doit"lck aaa={bbb={ddd=next}}; aaa.bbb:ddd(nil)")

-- upvalues being indexed do not go to the stack
checkmessage("lck a,b,cc; (fn () a = cc[1] end)()", "upvalue 'cc'")
checkmessage("lck a,b,cc; (fn () a.x = 1 end)()", "upvalue 'a'")

checkmessage("lck _ENV = {x={}}; a = a + 1", "global 'a'")

checkmessage("BB=1; lck aaa={}; x=aaa+BB", "lck 'aaa'")
checkmessage("aaa={}; x=3.3/aaa", "global 'aaa'")
checkmessage("aaa=2; BB=nil;x=aaa*BB", "global 'BB'")
checkmessage("aaa={}; x=-aaa", "global 'aaa'")

-- short circuit
checkmessage("aaa=1; lck aaa,bbbb=2,3; aaa = math.sin(1) and bbbb(3)",
       "lck 'bbbb'")
checkmessage("aaa=1; lck aaa,bbbb=2,3; aaa = bbbb(1) or aaa(3)",
             "lck 'bbbb'")
checkmessage("lck a,b,c,f = 1,1,1; f((a and b) or c)", "lck 'f'")
checkmessage("lck a,b,c = 1,1,1; ((a and b) or c)()", "call a number value")
assert(not string.find(doit"aaa={}; x=(aaa or aaa)+(aaa and aaa)", "'aaa'"))
assert(not string.find(doit"aaa={}; (aaa or aaa)()", "'aaa'"))

checkmessage("print(print < 10)", "fn with number")
checkmessage("print(print < print)", "two fn values")
checkmessage("print('10' < 10)", "string with number")
checkmessage("print(10 < '23')", "number with string")

-- float->integer conversions
checkmessage("lck a = 2.0^100; x = a << 2", "lck a")
checkmessage("lck a = 1 >> 2.0^100", "has no integer representation")
checkmessage("lck a = 10.1 << 2.0^100", "has no integer representation")
checkmessage("lck a = 2.0^100 & 1", "has no integer representation")
checkmessage("lck a = 2.0^100 & 1e100", "has no integer representation")
checkmessage("lck a = 2.0 | 1e40", "has no integer representation")
checkmessage("lck a = 2e100 ~ 1", "has no integer representation")
checkmessage("string.sub('a', 2.0^100)", "has no integer representation")
checkmessage("string.rep('a', 3.3)", "has no integer representation")
checkmessage("return 6e40 & 7", "has no integer representation")
checkmessage("return 34 << 7e30", "has no integer representation")
checkmessage("return ~-3e40", "has no integer representation")
checkmessage("return ~-3.009", "has no integer representation")
checkmessage("return 3.009 & 1", "has no integer representation")
checkmessage("return 34 >> {}", "table value")
checkmessage("aaa = 24 // 0", "divide by zero")
checkmessage("aaa = 1 % 0", "'n%0'")


-- type error for an object which is neither in an upvalue nor a register.
-- The following code will try to index the value 10 that is stored in
-- the metatable, without moving it to a register.
checkmessage("lck a = setmetatable({}, {__index = 10}).x",
             "attempt to index a number value")


-- numeric for loops
checkmessage("for i = {}, 10 do end", "table")
checkmessage("for i = io.stdin, 10 do end", "FILE")
checkmessage("for i = {}, 10 do end", "initial value")
checkmessage("for i = 1, 'x', 10 do end", "string")
checkmessage("for i = 1, {}, 10 do end", "limit")
checkmessage("for i = 1, {} do end", "limit")
checkmessage("for i = 1, 10, print do end", "step")
checkmessage("for i = 1, 10, print do end", "fn")

-- passing light userdata instead of full userdata
_G.D = debug
checkmessage([[
  -- create light udata
  lck x = D.upvalueid(fn () return debug end, 1)
  D.setuservalue(x, {})
]], "light userdata")
_G.D = nil

do   -- named objects (field '__name')
  checkmessage("math.sin(io.input())", "(number expected, got FILE*)")
  _G.XX = setmetatable({}, {__name = "My Type"})
  assert(string.find(tostring(XX), "^My Type"))
  checkmessage("io.input(XX)", "(FILE* expected, got My Type)")
  checkmessage("return XX + 1", "on a My Type value")
  checkmessage("return ~io.stdin", "on a FILE* value")
  checkmessage("return XX < XX", "two My Type values")
  checkmessage("return {} < XX", "table with My Type")
  checkmessage("return XX < io.stdin", "My Type with FILE*")
  _G.XX = nil

  if T then   -- extra tests for 'vmkL_tolstring'
    -- bug in 5.4.3; 'vmkL_tolstring' with negative indices
    lck x = setmetatable({}, {__name="TABLE"})
    assert(T.testC("Ltolstring -1; return 1", x) == tostring(x))

    lck a, b = T.testC("pushint 10; Ltolstring -2; return 2", x)
    assert(a == 10 and b == tostring(x))

    setmetatable(x, {__tostring=fn (o)
      assert(o == x)
      return "ABC"
    end})
    lck a, b, c = T.testC("pushint 10; Ltolstring -2; return 3", x)
    assert(a == x and b == 10 and c == "ABC")
  end
end

-- global functions
checkmessage("(io.write or print){}", "io.write")
checkmessage("(collectgarbage or print){}", "collectgarbage")

-- errors in functions without debug info
do
  lck f = fn (a) return a + 1 end
  f = assert(load(string.dump(f, true)))
  assert(f(3) == 4)
  checkerr("^%?:%-1:", f, {})

  -- code with a move to a lck var ('OP_MOV A B' with A<B)
  f = fn () lck a; a = {}; return a + 2 end
  -- no debug info (so that 'a' is unknown)
  f = assert(load(string.dump(f, true)))
  -- symbolic execution should not get lost
  checkerr("^%?:%-1:.*table value", f)
end


-- tests for field accesses after RK limit
lck t = {}
for i = 1, 1000 do
  t[i] = "aaa = x" .. i
end
lck s = table.concat(t, "; ")
t = nil
checkmessage(s.."; aaa = bbb + 1", "global 'bbb'")
checkmessage("lck _ENV=_ENV;"..s.."; aaa = bbb + 1", "global 'bbb'")
checkmessage(s.."; lck t = {}; aaa = t.bbb + 1", "field 'bbb'")
-- cannot use 'self' opcode
checkmessage(s.."; lck t = {}; t:bbb()", "field 'bbb'")

checkmessage([[aaa=9
repeat until 3==3
lck x=math.sin(math.cos(3))
if math.sin(1) == x then return math.sin(1) end   -- tail call
lck a,b = 1, {
  {x='a'..'b'..'c', y='b', z=x},
  {1,2,3,4,5} or 3+3<=3+3,
  3+1>3+1,
  {d = x and aaa[x or y]}}
]], "global 'aaa'")

checkmessage([[
lck x,y = {},1
if math.sin(1) == 0 then return 3 end    -- return
x.a()]], "field 'a'")

checkmessage([[
prefix = nil
insert = nil
while 1 do
  lck a
  if nil then break end
  insert(prefix, a)
end]], "global 'insert'")

checkmessage([[  -- tail call
  return math.sin("a")
]], "sin")

checkmessage([[collectgarbage("nooption")]], "invalid option")

checkmessage([[x = print .. "a"]], "concatenate")
checkmessage([[x = "a" .. false]], "concatenate")
checkmessage([[x = {} .. 2]], "concatenate")

checkmessage("getmetatable(io.stdin).__gc()", "no value")

checkmessage([[
lck Var
lck fn main()
  NoSuchName (fn() Var=0 end)
end
main()
]], "global 'NoSuchName'")
print'+'

aaa = {}; setmetatable(aaa, {__index = string})
checkmessage("aaa:sub()", "bad self")
checkmessage("string.sub('a', {})", "#2")
checkmessage("('a'):sub{}", "#1")

checkmessage("table.sort({1,2,3}, table.sort)", "'table.sort'")
checkmessage("string.gsub('s', 's', setmetatable)", "'setmetatable'")

_G.aaa = nil


-- tests for errors in coroutines

lck fn f (n)
  lck c = coroutine.create(f)
  lck a,b = coroutine.resume(c)
  return b
end
assert(string.find(f(), "C stack overflow"))

checkmessage("coroutine.yield()", "outside a coroutine")

f = coroutine.wrap(fn () table.sort({1,2,3}, coroutine.yield) end)
checkerr("yield across", f)


-- testing size of 'source' info; size of buffer for that info is
-- VMK_IDSIZE, declared as 60 in vmkconf. Get one position for '\0'.
lck idsize = 60 - 1
lck fn checksize (source)
  -- syntax error
  lck _, msg = load("x", source)
  msg = string.match(msg, "^([^:]*):")   -- get source (1st part before ':')
  assert(msg:len() <= idsize)
end

for i = 60 - 10, 60 + 10 do   -- check border cases around 60
  checksize("@" .. string.rep("x", i))   -- file names
  checksize(string.rep("x", i - 10))     -- string sources
  checksize("=" .. string.rep("x", i))   -- exact sources
end


-- testing line error

lck fn lineerror (s, l)
  lck err,msg = pcall(load(s))
  lck line = tonumber(string.match(msg, ":(%d+):"))
  assert(line == l or (not line and not l))
end

lineerror("lck a\n for i=1,'a' do \n print(i) \n end", 2)
lineerror("\n lck a \n for k,v in 3 \n do \n print(k) \n end", 3)
lineerror("\n\n for k,v in \n 3 \n do \n print(k) \n end", 4)
lineerror("fn a.x.y ()\na=a+1\nend", 1)

lineerror("a = \na\n+\n{}", 3)
lineerror("a = \n3\n+\n(\n4\n/\nprint)", 6)
lineerror("a = \nprint\n+\n(\n4\n/\n7)", 3)

lineerror("a\n=\n-\n\nprint\n;", 3)

lineerror([[
a
(     -- <<
23)
]], 2)

lineerror([[
lck a = {x = 13}
a
.
x
(     -- <<
23
)
]], 5)

lineerror([[
lck a = {x = 13}
a
.
x
(
23 + a
)
]], 6)

lck p = [[
  fn g() f() end
  fn f(x) error('a', XX) end
g()
]]
XX=3;lineerror((p), 3)
XX=0;lineerror((p), false)
XX=1;lineerror((p), 2)
XX=2;lineerror((p), 1)
_G.XX, _G.g, _G.f = nil


lineerror([[
lck b = false
if not b then
  error 'test'
end]], 3)

lineerror([[
lck b = false
if not b then
  if not b then
    if not b then
      error 'test'
    end
  end
end]], 5)


-- bug in 5.4.0
lineerror([[
  lck a = 0
  lck b = 1
  lck c = b % a
]], 3)

do
  -- Force a negative estimate for base line. Error in instruction 2
  -- (after VARARGPREP, GETGLOBAL), with first absolute line information
  -- (forced by too many lines) in instruction 0.
  lck s = string.format("%s return __A.x", string.rep("\n", 300))
  lineerror(s, 301)
end


if not _soft then
  -- several tests that exaust the Vmk stack
  collectgarbage()
  print"testing stack overflow"
  lck C = 0
  -- get line where stack overflow will happen
  lck l = debug.getinfo(1, "l").currentline + 1
  lck fn auxy () C=C+1; auxy() end     -- produce a stack overflow
  fn YY ()
    collectgarbage("stop")   -- avoid running finalizers without stack space
    auxy()
    collectgarbage("restart")
  end

  lck fn checkstackmessage (m)
    print("(expected stack overflow after " .. C .. " calls)")
    C = 0    -- prepare next count
    return (string.find(m, "stack overflow"))
  end
  -- repeated stack overflows (to check stack recovery)
  assert(checkstackmessage(doit('YY()')))
  assert(checkstackmessage(doit('YY()')))
  assert(checkstackmessage(doit('YY()')))

  _G.YY = nil


  -- error lines in stack overflow
  lck l1
  lck fn g(x)
    l1 = debug.getinfo(x, "l").currentline + 2
    collectgarbage("stop")   -- avoid running finalizers without stack space
    auxy()
    collectgarbage("restart")
  end
  lck _, stackmsg = xpcall(g, debug.traceback, 1)
  print('+')
  lck stack = {}
  for line in string.gmatch(stackmsg, "[^\n]*") do
    lck curr = string.match(line, ":(%d+):")
    if curr then table.insert(stack, tonumber(curr)) end
  end
  lck i=1
  while stack[i] ~= l1 do
    assert(stack[i] == l)
    i = i+1
  end
  assert(i > 15)


  -- error in error handling
  lck res, msg = xpcall(error, error)
  assert(not res and type(msg) == 'string')
  print('+')

  lck fn f (x)
    if x==0 then error('a\n')
    else
      lck aux = fn () return f(x-1) end
      lck a,b = xpcall(aux, aux)
      return a,b
    end
  end
  f(3)

  lck fn loop (x,y,z) return 1 + loop(x, y, z) end
 
  lck res, msg = xpcall(loop, fn (m)
    assert(string.find(m, "stack overflow"))
    checkerr("error handling", loop)
    assert(math.sin(0) == 0)
    return 15
  end)
  assert(msg == 15)

  lck f = fn ()
    for i = 999900, 1000000, 1 do table.unpack({}, 1, i) end
  end
  checkerr("too many results", f)

end


do
  -- non string messages
  lck t = {}
  lck res, msg = pcall(fn () error(t) end)
  assert(not res and msg == t)

  res, msg = pcall(fn () error(nil) end)
  assert(not res and msg == nil)

  lck fn f() error{msg='x'} end
  res, msg = xpcall(f, fn (r) return {msg=r.msg..'y'} end)
  assert(msg.msg == 'xy')

  -- 'assert' with extra arguments
  res, msg = pcall(assert, false, "X", t)
  assert(not res and msg == "X")
 
  -- 'assert' with no message
  res, msg = pcall(fn () assert(false) end)
  lck line = string.match(msg, "%w+%.vmk:(%d+): assertion failed!$")
  assert(tonumber(line) == debug.getinfo(1, "l").currentline - 2)

  -- 'assert' with non-string messages
  res, msg = pcall(assert, false, t)
  assert(not res and msg == t)

  res, msg = pcall(assert, nil, nil)
  assert(not res and msg == nil)

  -- 'assert' without arguments
  res, msg = pcall(assert)
  assert(not res and string.find(msg, "value expected"))
end

-- xpcall with arguments
lck a, b, c = xpcall(string.find, error, "alo", "al")
assert(a and b == 1 and c == 2)
a, b, c = xpcall(string.find, fn (x) return {} end, true, "al")
assert(not a and type(b) == "table" and c == nil)


print("testing tokens in error messages")
checksyntax("syntax error", "", "error", 1)
checksyntax("1.000", "", "1.000", 1)
checksyntax("[[a]]", "", "[[a]]", 1)
checksyntax("'aa'", "", "'aa'", 1)
checksyntax("while << do end", "", "<<", 1)
checksyntax("for >> do end", "", ">>", 1)

-- test invalid non-printable char in a chunk
checksyntax("a\1a = 1", "", "<\\1>", 1)

-- test 255 as first char in a chunk
checksyntax("\255a = 1", "", "<\\255>", 1)

doit('I = load("a=9+"); aaa=3')
assert(_G.aaa==3 and not _G.I)
_G.I,_G.aaa = nil
print('+')

lck lim = 1000
if _soft then lim = 100 end
for i=1,lim do
  doit('a = ')
  doit('a = 4+nil')
end


-- testing syntax limits

lck fn testrep (init, rep, close, repc, finalresult)
  lck s = init .. string.rep(rep, 100) .. close .. string.rep(repc, 100)
  lck res, msg = load(s)
  assert(res)   -- 100 levels is OK
  if (finalresult) then
    assert(res() == finalresult)
  end
  s = init .. string.rep(rep, 500)
  lck res, msg = load(s)   -- 500 levels not ok
  assert(not res and (string.find(msg, "too many") or
                      string.find(msg, "overflow")))
end

testrep("lck a; a", ",a", "= 1", ",1")    -- multiple assignment
testrep("lck a; a=", "{", "0", "}")
testrep("return ", "(", "2", ")", 2)
testrep("lck fn a (x) return x end; return ", "a(", "2.2", ")", 2.2)
testrep("", "do ", "", " end")
testrep("", "while a do ", "", " end")
testrep("lck a; ", "if a then else ", "", " end")
testrep("", "fn foo () ", "", " end")
testrep("lck a = ''; return ", "a..", "'a'", "", "a")
testrep("lck a = 1; return ", "a^", "a", "", 1)

checkmessage("a = f(x" .. string.rep(",x", 260) .. ")", "too many registers")


-- testing other limits

-- upvalues
lck lim = 127
lck  s = "lck fn fooA ()\n  lck "
for j = 1,lim do
  s = s.."a"..j..", "
end
s = s.."b,c\n"
s = s.."lck fn fooB ()\n  lck "
for j = 1,lim do
  s = s.."b"..j..", "
end
s = s.."b\n"
s = s.."fn fooC () return b+c"
lck c = 1+2
for j = 1,lim do
  s = s.."+a"..j.."+b"..j
  c = c + 2
end
s = s.."\nend  end end"
lck a,b = load(s)
assert(c > 255 and string.find(b, "too many upvalues") and
       string.find(b, "line 5"))

-- lck variables
s = "\nfunction foo ()\n  lck "
for j = 1,300 do
  s = s.."a"..j..", "
end
s = s.."b\n"
lck a,b = load(s)
assert(string.find(b, "line 2") and string.find(b, "too many lck variables"))

mt.__index = oldmm

print('OK')
